ludwig/schema/config_object.py

```diff
@@ -0,0 +1,469 @@
+import copy
+import sys
+import warnings
+from dataclasses import dataclass
+from typing import Dict, List
+
+import yaml
+from marshmallow import ValidationError
+
+from ludwig.constants import (
+    ACTIVE,
+    BINARY,
+    CATEGORY,
+    COLUMN,
+    COMBINER,
+    DECODER,
+    DEFAULT_VALIDATION_METRIC,
+    DEFAULTS,
+    ENCODER,
+    EXECUTOR,
+    HYPEROPT,
+    INPUT_FEATURES,
+    LOSS,
+    MODEL_ECD,
+    MODEL_GBM,
+    MODEL_TYPE,
+    NAME,
+    NUMBER,
+    OPTIMIZER,
+    OUTPUT_FEATURES,
+    PREPROCESSING,
+    PROC_COLUMN,
+    RAY,
+    SEQUENCE,
+    SPLIT,
+    TIED,
+    TRAINER,
+    TYPE,
+)
+from ludwig.features.feature_utils import compute_feature_hash
+from ludwig.modules.loss_modules import get_loss_cls
+from ludwig.schema import validate_config
+from ludwig.schema.combiners.base import BaseCombinerConfig
+from ludwig.schema.combiners.concat import ConcatCombinerConfig
+from ludwig.schema.combiners.utils import combiner_registry
+from ludwig.schema.decoders.utils import get_decoder_cls
+from ludwig.schema.defaults.defaults import DefaultsConfig
+from ludwig.schema.encoders.base import PassthroughEncoderConfig
+from ludwig.schema.encoders.binary_encoders import BinaryPassthroughEncoderConfig
+from ludwig.schema.encoders.utils import get_encoder_cls
+from ludwig.schema.features.base import BaseFeatureConfig
+from ludwig.schema.features.utils import get_input_feature_cls, get_output_feature_cls, input_config_registry
+from ludwig.schema.optimizers import get_optimizer_cls
+from ludwig.schema.preprocessing import PreprocessingConfig
+from ludwig.schema.split import get_split_cls
+from ludwig.schema.trainer import BaseTrainerConfig, ECDTrainerConfig, GBMTrainerConfig
+from ludwig.schema.utils import BaseMarshmallowConfig, convert_submodules
+from ludwig.utils.backward_compatibility import upgrade_to_latest_version
+from ludwig.utils.misc_utils import set_default_value
+
+DEFAULTS_MODULES = {NAME, COLUMN, PROC_COLUMN, TYPE, TIED, DEFAULT_VALIDATION_METRIC}
+
+
+class BaseFeatureContainer:
+    """Base Feature container for input and output features."""
+
+    def to_dict(self):
+        """Method for getting a dictionary representation of the input features.
+
+        Returns:
+            Dictionary of input features specified.
+        """
+        return convert_submodules(self.__dict__)
+
+    def to_list(self):
+        """Method for getting a list representation of the input features.
+
+        Returns:
+            List of input features specified.
+        """
+        return list(convert_submodules(self.__dict__).values())
+
+    def filter_features(self):
+        return {
+            key: {k: v for k, v in value.items() if k in {NAME, TYPE, ACTIVE}} for key, value in self.to_dict().items()
+        }
+
+    def __repr__(self):
```

Comment:
I'm probably being overly pedantic, but I think this is the intended use of `__str__`, not `__repr__`.

Reply:
So I think that in this case we actually do need to use `__repr__` because the user is interacting with the object itself in the sdk. So when the user calls `config` in a notebook cell using the SDK, the intent of defining these custom `__repr__` functions is so that they can see a good representation of the config. With `__str__` the user would have to surround each call to any section of the config object in a print statement which is ok, but I feel like it's a better UX to just be able to call the object directly and get a user friendly representation without having to surround it in a print each time. WDYT?

Reply:
Yes you're absolutely right, however, I think this is a slightly worse UX for the SDK user since they will have to surround each section of the config object in a print statement when they want to view it, and then remove the print statement when they want to set it. Alternatively if we use a `__repr__`, they can just call the section to view it and subsequently set it much easier. Here is an example of the two flows side by side.

```
config.input_features.Age.preprocessing
```
missing_value_strategy: fill_with_const
fill_value: 0.0
computed_fill_value: 0.0
normalization: null
```
config.input_features.Age.preprocessing.fill_value = 2.0
```

VS. 


```
print(config.input_features.Age.preprocessing)
```
missing_value_strategy: fill_with_const
fill_value: 0.0
computed_fill_value: 0.0
normalization: null
```
config.input_features.Age.preprocessing.fill_value = 2.0
```

I know it may seem like a small thing, however, after testing out the config object in a notebook setting, it is clear that the users will be traversing the config object quite a bit to see the parameters that are set and having to add and remove print statements throughout this process will be quite tedious.